src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -0,0 +1,1062 @@
+  // than _regular_group_size because no group is allowed to
+  // have smaller assignments than _smallest_chunk_size, which is 128 KB.
+
+  // Under normal circumstances, no configuration needs more than _maximum_groups (default value of 16).
+  // The first group "effectively" processes chunks of size 1 MiB (or smaller for smaller region sizes).
+  // The last group processes chunks of size 128 KiB.  There are four groups total.
+
+  // group[0] is 4 MiB chunk size (_maximum_chunk_size_words)
+  // group[1] is 2 MiB chunk size
+  // group[2] is 1 MiB chunk size
+  // group[3] is 512 KiB chunk size
+  // group[4] is 256 KiB chunk size
+  // group[5] is 128 Kib shunk size (_smallest_chunk_size_words = 4 * 64 * 64
+  static const size_t _maximum_groups = 6;
+
+  const ShenandoahHeap* _heap;
+
+  const size_t _regular_group_size;                        // Number of chunks in each group
+  const size_t _first_group_chunk_size_b4_rebalance;
+  const size_t _num_groups;                        // Number of groups in this configuration
+  const size_t _total_chunks;
+
+  shenandoah_padding(0);
+  volatile size_t _index;
+  shenandoah_padding(1);
+
+  size_t _region_index[_maximum_groups];           // The region index for the first region spanned by this group
+  size_t _group_offset[_maximum_groups];           // The offset at which group begins within first region spanned by this group
+  size_t _group_chunk_size[_maximum_groups];       // The size of each chunk within this group
+  size_t _group_entries[_maximum_groups];          // Total chunks spanned by this group and the ones before it.
+
+  // No implicit copying: iterators should be passed by reference to capture the state
+  NONCOPYABLE(ShenandoahRegionChunkIterator);
+
+  // Makes use of _heap.
+  size_t calc_regular_group_size();
+
+  // Makes use of _regular_group_size, which must be initialized before call.
+  size_t calc_first_group_chunk_size_b4_rebalance();
+
+  // Makes use of _regular_group_size and _first_group_chunk_size_b4_rebalance, both of which must be initialized before call.
+  size_t calc_num_groups();
+
+  // Makes use of _regular_group_size, _first_group_chunk_size_b4_rebalance, which must be initialized before call.
+  size_t calc_total_chunks();
+
+public:
+  ShenandoahRegionChunkIterator(size_t worker_count);
+  ShenandoahRegionChunkIterator(ShenandoahHeap* heap, size_t worker_count);
+
+  // Reset iterator to default state
+  void reset();
+
+  // Fills in assignment with next chunk of work and returns true iff there is more work.
+  // Otherwise, returns false.  This is multi-thread-safe.
+  inline bool next(struct ShenandoahRegionChunk *assignment);
+
+  // This is *not* MT safe. However, in the absence of multithreaded access, it
+  // can be used to determine if there is more work to do.
+  inline bool has_next() const;
+};
+
+typedef ShenandoahScanRemembered<ShenandoahDirectCardMarkRememberedSet> RememberedScanner;
```

Comment:
Do I see this correctly that there is only one implementation of the template parameter that is passed around all over this and related source files? If yes, then we could just as well clean this up, remove all the templating and move much/all of the code out of .inline.hpp and into .cpp file instead, and avoid making it inline, which might be counter-productive.

Reply:
Tracking this work in ticket https://bugs.openjdk.org/browse/JDK-8338479

Reply:
https://github.com/openjdk/shenandoah/pull/472